lms/djangoapps/courseware/courses.py

```diff
@@ -566,10 +566,18 @@ def get_course_blocks_completion_summary(course_key, user):
     course_usage_key = store.make_course_usage_key(course_key)
     block_data = get_course_blocks(user, course_usage_key, allow_start_dates_in_future=True, include_completion=True)
 
+    def _is_optional(*keys):
+        for key in keys:
+            if block_data.get_xblock_field(key, 'optional_content', False):
+                return True
+        return False
+
     complete_count, incomplete_count, locked_count = 0, 0, 0
     for section_key in block_data.get_children(course_usage_key):  # pylint: disable=too-many-nested-blocks
         for subsection_key in block_data.get_children(section_key):
             for unit_key in block_data.get_children(subsection_key):
+                if optional != _is_optional(section_key, subsection_key, unit_key):
+                    continue
```

Comment:
What if the `section` sets the value of the `optional_content` field to `True` but the `subsection` sets it to `False`? Would it make sense to disallow marking `subsection` as non-optional when its parent `section` is marked as optional?

Reply:
I was handling it as if a parent was optional all children would behave as optional regardless of their actual value, but now that you mention it, in order to reduce confusion it'd be better to disable the checkbox when a parent is optional.